### PR TITLE
Fix trees

### PR DESCRIPTION
**Pull request by nyee**
Some changes related to my attempts to fix the trees as related to RMG-database issue

### CODE REVIEW

**Comment by rwest** [comment]
I made some suggestions. Once you've checked them, @nyee, you should be able to
`git rebase -i --autosquash master fixTrees`  and it will put them into place as fixup commits and merge them with yours.

**Comment by nyee** [comment]
Thanks for the suggestions, Richard. I have merged them in and forced pushed the updated version to this branch.

**Comment by nyee** [comment]
This latest test passes for the RMG-database branch [fixTrees](https://github.com/nyee/RMG-database/tree/fixTrees).  The same test should be implemented on kinetics, solvation, transport, and statmech also, but I didn't want to put it in yet because no RMG-database commit will pass those tests.

In terms of debugging those other databases: It is easier to use this script to automatically rearrange the trees:

[fixSiblings.txt](https://github.com/ReactionMechanismGenerator/RMG-Py/files/298335/fixSiblings.txt)

The above script should **not** be trusted used blindly. For example, it might be that groups are not [explicit enough](https://github.com/nyee/RMG-database/commit/71ea6851397cfb808a41ea7e6812a1e50ca1b7b4), rather than the tree is set up wrong.

File: rmgpy/data/base.py [hunk 1]
```diff
@@ -875,10 +875,10 @@ def matchNodeToChild(self, parentNode, childNode):
             return False
         
         #If the parentNode is a Group and the childNode is a LogicOr there is nothing to check,
-        #so it gets an automatic pass. However, we do need to check that everything down this
+        #except that the parent is listed in the attributes. However, we do need to check that everything down this
         #family line is consistent, which is done in the databaseTest unitTest
         elif isinstance(parentNode.item, Group) and isinstance(childNode.item, LogicOr):
-            return True
+            return childNode.parent is parentNode
```
Comment (KEHANG) [inline rmgpy/data/base.py]: in order to spot possible suspicious parent/child, for this if situation `isinstance(parentNode.item, Group) and isinstance(childNode.item, LogicOr)`, we  might also want to check the items in the `LogicOr` are children of the `parentNode`

File: rmgpy/data/thermo.py [hunk 2]
```diff
@@ -423,6 +423,53 @@ def processOldLibraryEntry(self, data):
...
+        #Remove from parent's children attribute
+        parentR.children.remove(entryToRemove)
+
+        #change children's parent attribute to former grandparent
+        for child in entryToRemove.children:
+            child.parent = parentR
+
+        #extend parent's children attribute with new children
+        parentR.children.extend(entryToRemove.children)
+
```
Comment (KEHANG) [inline rmgpy/data/thermo.py]: please add an unittest for this helpful method

File: testing/databaseTest.py [hunk 3]
```diff
@@ -259,6 +277,26 @@ def kinetics_checkChildParentRelationships(self, family_name):
                     nose.tools.assert_true(family.matchNodeToChild(ancestorNode, childNode),
                                     "In {family} family, group {ancestor} is not a proper ancestor of its child {child}.".format(family=family_name, ancestor=ancestorNode, child=nodeName))
 
+    def kinetics__checkSiblingsForParents(self, family_name):
```
Comment (KEHANG) [inline testing/databaseTest.py]: remove one underscore in the method name?

File: rmgpy/data/kinetics/common.py [hunk 4]
```diff
@@ -83,40 +83,44 @@ def sortEfficiencies(efficiencies0):
     if entry.label != '':
         f.write('    label = "{0}",\n'.format(entry.label))
 
+
+
     if isinstance(entry.item, Reaction):
-        if entry.label != str(entry.item):
```
Comment (KEHANG) [inline rmgpy/data/kinetics/common.py]: seems this `saveEntry` method is used for many purposes, can we also add some comments in the method to clarify which case (e.g., save to `groups.py`, `rules.py`, `reactions.py`) is for which if statement?

File: testing/databaseTest.py (hunk 3 above)
Comment (nyee) [inline testing/databaseTest.py]: Did some rebasing and this is now fixed in commit 1bba6c

File: rmgpy/data/kinetics/common.py (hunk 4 above)
Comment (nyee) [inline rmgpy/data/kinetics/common.py]: Fixed up in commit fdb29

**Comment by nyee** [comment]
Added unittest for removeGroup with latest commit. Discovered some bugs along the way and fixed them in the fixup! commit for the removeGroup method.

Unfortunately, I can't run make test on my computer, and I'm not sure I used the correct path to the test_data, so I need travis to tell me whether the unittest passes.

File: rmgpy/data/base.py (hunk 1 above)
Comment (nyee) [inline rmgpy/data/base.py]: As said in the comment above, we actually do this in [databaseTest.py](https://github.com/ReactionMechanismGenerator/RMG-Py/blob/059b895b67370b2daa5699c3b5c84ffc87a36add/testing/databaseTest.py#L245-L253
l). 

In my opinion, the current implementation, unless we want to start using the matchNodeToChild for things other than database unittests.

File: rmgpy/data/thermo.py [hunk 5]
```diff
@@ -421,6 +421,66 @@ def processOldLibraryEntry(self, data):
...
+                parentR.reference = groupToRemove.reference
+                parentR._longDesc = groupToRemove._longDesc
+                parentR._shortDesc = groupToRemove._shortDesc
+                parentR._longDesc = groupToRemove.longDesc
+                parentR._shortDesc = groupToRemove.shortDesc
+                parentR.rank = groupToRemove.rank
+                parentR.referenceType = groupToRemove.referenceType
+
+        #look for other pointers that point toward entry
+        for entryName, entry in self.entries.iteritems():
```
Comment (KEHANG) [inline rmgpy/data/thermo.py]: I think this `for` should be combined with the `if` statement starting at line 457 in order to make it more concise.

**Pull request merged**